File: app/datatables/student_datatable.rb
class StudentDatatable < AjaxDatatablesRails::ActiveRecord
  def view_columns
    @view_columns ||= {
      admission_no: { source: 'Student.admission_no' },
      name: { source: 'Student.name' },
      date_of_birth: { source: 'Student.date_of_birth' },
      mobile_number: { source: 'Student.mobile_number', searchable: true },
      standard_name: { source: 'Standard.name', searchable: true },
      section_name: { source: 'Section.section_name', searchable: true },
      roll_no: { source: 'Student.roll_no', searchable: true },
      gender: { source: 'Student.gender', searchable: false },
      student_status: { source: 'StudentDecorator.student_status', searchable: false },
      student_actions: { source: 'StudentDecorator.student_actions', searchable: false }
    }
  end

  def data
    records.map do |record|
      {
        admission_no: record.decorate.link_to,
        name: record.name,
        date_of_birth: record.date_of_birth&.strftime('%d-%m-%Y'),
        mobile_number: record.mobile_number,
        standard_name: record.section.standard.name,
        section_name: record.section.section_name,
        roll_no: record.roll_no,
        gender: record.gender,
        student_status: record.decorate.student_status,
        student_actions: record.decorate.student_actions
      }
    end
  end

  def get_raw_records
    Student.joins(section: :standard).where(removed: false)
  end
end

File: app/javascript/controllers/student/student_datatable.js
function initValidate() {
  $("#studentTable").dataTable({
    processing: true,
    serverSide: true,
    ajax: {
      url: $("#studentTable").data("source"),
    },
    pagingType: "simple_numbers",
    language: {
      sLengthMenu: "_MENU_ records per page",
    },
    stateSave: false,
    columns: [
      { data: "admission_no" },
      { data: "name" },
      { data: "date_of_birth" },
      { data: "mobile_number"},
      { data: "standard_name"},
      { data: "section_name"},
      { data: "roll_no" },
      { data: "gender" },
      { data: "student_status" },
      { data: "student_actions", class: "crud-action" },
    ],
    scrollX: "100%", // Enable horizontal scrolling
    oLanguage: {
      oPaginate: {
        sPrevious: '<span class="pagination-btn"><i class="fas fa-arrow-left"></i> Previous</span>',
        sNext: '<span class="pagination-btn">Next <i class="fas fa-arrow-right"></i></span>',
      },
    },
  });
}

$(document).ready(function () {
  initValidate();

  $(document).on("turbo:render", function () {
    initValidate();
  });
});
